File: inventory/pos/report/pos_profit_analysis/pos_profit_analysis.js
// Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["POS Profit Analysis"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "pos_session",
			"label": __("POS Session"),
			"fieldtype": "Link",
			"options": "POS Session"
		},
		{
			"fieldname": "item_code",
			"label": __("Item"),
			"fieldtype": "Link",
			"options": "Item"
		},
		{
			"fieldname": "customer",
			"label": __("Customer"),
			"fieldtype": "Data"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Highlight negative profit in red
		if (column.fieldname === "profit_amount" && data && data.profit_amount < 0) {
			value = `<span style="color: red">${value}</span>`;
		}
		
		// Highlight low profit margin in orange
		if (column.fieldname === "profit_margin_percent" && data && data.profit_margin_percent < 10) {
			value = `<span style="color: orange">${value}</span>`;
		}
		
		// Highlight high profit margin in green
		if (column.fieldname === "profit_margin_percent" && data && data.profit_margin_percent > 30) {
			value = `<span style="color: green">${value}</span>`;
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Add custom buttons
		report.page.add_inner_button(__("Export to Excel"), function() {
			frappe.utils.make_event_emitter(frappe.query_report);
			return frappe.query_report.export_report();
		});
		
		report.page.add_inner_button(__("Item Wise Summary"), function() {
			show_item_wise_summary(report);
		});
		
		report.page.add_inner_button(__("Session Wise Summary"), function() {
			show_session_wise_summary(report);
		});
	}
};

function show_item_wise_summary(report) {
	let data = report.data;
	if (!data || data.length === 0) {
		frappe.msgprint(__("No data to summarize"));
		return;
	}
	
	// Group by item
	let item_summary = {};
	data.forEach(row => {
		if (!item_summary[row.item_code]) {
			item_summary[row.item_code] = {
				item_name: row.item_name,
				total_qty: 0,
				total_sales: 0,
				total_cost: 0,
				total_profit: 0
			};
		}
		
		item_summary[row.item_code].total_qty += row.qty;
		item_summary[row.item_code].total_sales += row.amount;
		item_summary[row.item_code].total_cost += row.total_cost;
		item_summary[row.item_code].total_profit += row.profit_amount;
	});
	
	// Create summary table
	let summary_html = `
		<table class="table table-bordered">
			<thead>
				<tr>
					<th>Item Code</th>
					<th>Item Name</th>
					<th>Total Qty</th>
					<th>Total Sales</th>
					<th>Total Cost</th>
					<th>Total Profit</th>
					<th>Profit %</th>
				</tr>
			</thead>
			<tbody>
	`;
	
	Object.keys(item_summary).forEach(item_code => {
		let item = item_summary[item_code];
		let profit_percent = item.total_sales > 0 ? (item.total_profit / item.total_sales * 100).toFixed(2) : 0;
		
		summary_html += `
			<tr>
				<td>${item_code}</td>
				<td>${item.item_name}</td>
				<td>${item.total_qty.toFixed(2)}</td>
				<td>${format_currency(item.total_sales)}</td>
				<td>${format_currency(item.total_cost)}</td>
				<td>${format_currency(item.total_profit)}</td>
				<td>${profit_percent}%</td>
			</tr>
		`;
	});
	
	summary_html += `
			</tbody>
		</table>
	`;
	
	frappe.msgprint({
		title: __("Item Wise Profit Summary"),
		message: summary_html,
		wide: true
	});
}

function show_session_wise_summary(report) {
	let data = report.data;
	if (!data || data.length === 0) {
		frappe.msgprint(__("No data to summarize"));
		return;
	}
	
	// Group by session
	let session_summary = {};
	data.forEach(row => {
		if (!session_summary[row.pos_session]) {
			session_summary[row.pos_session] = {
				date: row.posting_date,
				total_qty: 0,
				total_sales: 0,
				total_cost: 0,
				total_profit: 0,
				invoice_count: new Set()
			};
		}
		
		session_summary[row.pos_session].total_qty += row.qty;
		session_summary[row.pos_session].total_sales += row.amount;
		session_summary[row.pos_session].total_cost += row.total_cost;
		session_summary[row.pos_session].total_profit += row.profit_amount;
		session_summary[row.pos_session].invoice_count.add(row.invoice_name);
	});
	
	// Create summary table
	let summary_html = `
		<table class="table table-bordered">
			<thead>
				<tr>
					<th>Session</th>
					<th>Date</th>
					<th>Invoices</th>
					<th>Total Qty</th>
					<th>Total Sales</th>
					<th>Total Cost</th>
					<th>Total Profit</th>
					<th>Profit %</th>
				</tr>
			</thead>
			<tbody>
	`;
	
	Object.keys(session_summary).forEach(session => {
		let sess = session_summary[session];
		let profit_percent = sess.total_sales > 0 ? (sess.total_profit / sess.total_sales * 100).toFixed(2) : 0;
		
		summary_html += `
			<tr>
				<td>${session}</td>
				<td>${frappe.datetime.str_to_user(sess.date)}</td>
				<td>${sess.invoice_count.size}</td>
				<td>${sess.total_qty.toFixed(2)}</td>
				<td>${format_currency(sess.total_sales)}</td>
				<td>${format_currency(sess.total_cost)}</td>
				<td>${format_currency(sess.total_profit)}</td>
				<td>${profit_percent}%</td>
			</tr>
		`;
	});
	
	summary_html += `
			</tbody>
		</table>
	`;
	
	frappe.msgprint({
		title: __("Session Wise Profit Summary"),
		message: summary_html,
		wide: true
	});
}

File: inventory/pos/report/pos_profit_analysis/pos_profit_analysis.py
# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate


def execute(filters=None):
	"""Main report execution function"""
	if not filters:
		filters = {}
	
	columns = get_columns()
	data = get_data(filters)
	chart = get_chart_data(data)
	summary = get_summary(data)
	
	return columns, data, None, chart, summary


def get_columns():
	"""Define report columns"""
	return [
		{
			"fieldname": "posting_date",
			"label": _("Date"),
			"fieldtype": "Date",
			"width": 100
		},
		{
			"fieldname": "pos_session",
			"label": _("POS Session"),
			"fieldtype": "Link",
			"options": "POS Session",
			"width": 150
		},
		{
			"fieldname": "invoice_name",
			"label": _("Invoice"),
			"fieldtype": "Link",
			"options": "POS Invoice",
			"width": 120
		},
		{
			"fieldname": "customer",
			"label": _("Customer"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "item_code",
			"label": _("Item Code"),
			"fieldtype": "Link",
			"options": "Item",
			"width": 120
		},
		{
			"fieldname": "item_name",
			"label": _("Item Name"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "qty",
			"label": _("Qty"),
			"fieldtype": "Float",
			"width": 80
		},
		{
			"fieldname": "rate",
			"label": _("Rate"),
			"fieldtype": "Currency",
			"width": 100
		},
		{
			"fieldname": "amount",
			"label": _("Amount"),
			"fieldtype": "Currency",
			"width": 100
		},
		{
			"fieldname": "cost_price",
			"label": _("Cost Price"),
			"fieldtype": "Currency",
			"width": 100
		},
		{
			"fieldname": "total_cost",
			"label": _("Total Cost"),
			"fieldtype": "Currency",
			"width": 100
		},
		{
			"fieldname": "profit_amount",
			"label": _("Profit"),
			"fieldtype": "Currency",
			"width": 100
		},
		{
			"fieldname": "profit_margin_percent",
			"label": _("Profit %"),
			"fieldtype": "Percent",
			"width": 100
		}
	]


def get_data(filters):
	"""Get report data based on filters"""
	conditions = get_conditions(filters)
	
	query = f"""
		SELECT 
			pi.posting_date,
			pi.pos_session,
			pi.name as invoice_name,
			pi.customer,
			pii.item_code,
			pii.item_name,
			pii.qty,
			pii.rate,
			pii.amount,
			pii.cost_price,
			(pii.cost_price * pii.qty) as total_cost,
			pii.profit_amount,
			pii.profit_margin_percent
		FROM `tabPOS Invoice` pi
		INNER JOIN `tabPOS Invoice Item` pii ON pi.name = pii.parent
		WHERE pi.docstatus = 1 {conditions}
		ORDER BY pi.posting_date DESC, pi.name, pii.idx
	"""
	
	return frappe.db.sql(query, filters, as_dict=1)


def get_conditions(filters):
	"""Build SQL conditions based on filters"""
	conditions = []
	
	if filters.get("from_date"):
		conditions.append("pi.posting_date >= %(from_date)s")
	
	if filters.get("to_date"):
		conditions.append("pi.posting_date <= %(to_date)s")
	
	if filters.get("pos_session"):
		conditions.append("pi.pos_session = %(pos_session)s")
	
	if filters.get("item_code"):
		conditions.append("pii.item_code = %(item_code)s")
	
	if filters.get("customer"):
		conditions.append("pi.customer = %(customer)s")
	
	return " AND " + " AND ".join(conditions) if conditions else ""


def get_chart_data(data):
	"""Generate chart data for profit analysis"""
	if not data:
		return None
	
	# Group data by date for chart
	date_wise_data = {}
	for row in data:
		date = row.posting_date
		if date not in date_wise_data:
			date_wise_data[date] = {
				"sales": 0,
				"cost": 0,
				"profit": 0
			}
		
		date_wise_data[date]["sales"] += flt(row.amount)
		date_wise_data[date]["cost"] += flt(row.total_cost)
		date_wise_data[date]["profit"] += flt(row.profit_amount)
	
	# Sort dates
	sorted_dates = sorted(date_wise_data.keys())
	
	return {
		"data": {
			"labels": [formatdate(date) for date in sorted_dates],
			"datasets": [
				{
					"name": "Sales",
					"values": [date_wise_data[date]["sales"] for date in sorted_dates]
				},
				{
					"name": "Cost",
					"values": [date_wise_data[date]["cost"] for date in sorted_dates]
				},
				{
					"name": "Profit",
					"values": [date_wise_data[date]["profit"] for date in sorted_dates]
				}
			]
		},
		"type": "line",
		"height": 300
	}


def get_summary(data):
	"""Generate summary statistics"""
	if not data:
		return []
	
	total_sales = sum(flt(row.amount) for row in data)
	total_cost = sum(flt(row.total_cost) for row in data)
	total_profit = sum(flt(row.profit_amount) for row in data)
	total_qty = sum(flt(row.qty) for row in data)
	
	overall_margin = (total_profit / total_sales * 100) if total_sales > 0 else 0
	
	return [
		{
			"value": total_sales,
			"label": _("Total Sales"),
			"datatype": "Currency"
		},
		{
			"value": total_cost,
			"label": _("Total Cost"),
			"datatype": "Currency"
		},
		{
			"value": total_profit,
			"label": _("Total Profit"),
			"datatype": "Currency"
		},
		{
			"value": overall_margin,
			"label": _("Overall Margin %"),
			"datatype": "Percent"
		},
		{
			"value": total_qty,
			"label": _("Total Quantity"),
			"datatype": "Float"
		},
		{
			"value": len(data),
			"label": _("Total Items Sold"),
			"datatype": "Int"
		}
	]